### PR TITLE
ncm-opennebula: Sunstone configuration and shared datastores support

### LINKED ISSUE #352: ncm-opennebula: ~/.one/one_auth file is not replaced by ~/.one/one_auth.new
After component execution the file ~/.one/one_auth is not overwritten by Quattor.
The new oneadmin password is stored in  ~/.one/one_auth.new instead ~/.one/one_auth

### LINKED ISSUE #355: ncm-opennebula: sunstone service passwd is not set by quattor
Set sunstone password from quattor as well

### LINKED ISSUE #493: ncm-opennebula: Include support for shared datastores
At this moment only Ceph datastores are supported by ONE component.

### LINKED ISSUE #494: ncm-opennebula: Configure sunstone-server.conf from one component
At this moment sunstone-server.conf is not configured by the component

### PR DESCRIPTION
**Pull request by alvarosimon**
This PR includes several bug fixes:
 * Fixes #352 : ~/.one/one_auth file is not replaced by ~/.one/one_auth.new
 * Fixes #355 : Sunstone service passwd is not set by quattor
 * Fixes #493 : Include support for shared datastores
 * Fixes #494 : Configure sunstone-server.conf from one component

It also includes new Showback support to show resource usage cost reports: http://docs.opennebula.org/4.12/administration/users_and_groups/showback.html#showback

### CODE REVIEW

**Comment by stdweird** [comment]
@samary @StephaneGerardVUB can you test the nfs datastore?

**Comment by alvarosimon** [comment]
@stdweird meanwhile I'm testing the new sunstone setup in altaria

File: ncm-opennebula/src/main/pan/components/opennebula/schema.pan [hunk 1]
```diff
@@ -258,12 +349,13 @@ datastores, vnets, hosts names, etc
...
     include structure_component
     'datastores'    : opennebula_datastore[1..]
-    'users'         : opennebula_user[]
+    'users'         ? opennebula_user[]
     'vnets'         : opennebula_vnet[]
     'hosts'         : string[]
     'rpc'           : opennebula_rpc
     'untouchables'  : opennebula_untouchables
     'oned'          : opennebula_oned
+    'sunstone'      : opennebula_sunstone
```
Comment (stdweird) [inline ncm-opennebula/src/main/pan/components/opennebula/schema.pan]: is it really mandatory?

File: ncm-opennebula/src/main/perl/opennebula.pm [hunk 2]
```diff
@@ -20,6 +20,11 @@ use Readonly;
 
 Readonly::Scalar my $CEPHSECRETFILE => "/var/lib/one/templates/secret/secret_ceph.xml";
 Readonly::Scalar our $ONED_CONF_FILE => "/etc/one/oned.conf";
+Readonly::Scalar our $SUNSTONE_CONF_FILE => "/etc/one/sunstone-server.conf";
+Readonly::Scalar our $ONEADMIN_AUTH_FILE => "/var/lib/one/.one/one_auth";
+Readonly::Scalar our $SERVERADMIN_AUTH_DIR => "/var/lib/one/.one/";
+Readonly::Array our @SERVERADMIN_AUTH_FILE => ("sunstone_auth", "oneflow_auth",
```
Comment (stdweird) [inline ncm-opennebula/src/main/perl/opennebula.pm]: use `qw()`

File: ncm-opennebula/src/main/perl/opennebula.pm [hunk 3]
```diff
@@ -467,6 +479,9 @@ sub manage_users
 
     foreach my $user (@$users) {
         if ($user->{user}) {
+            if ($user->{user} eq "serveradmin" && exists $user->{password}) {
```
Comment (stdweird) [inline ncm-opennebula/src/main/perl/opennebula.pm]: why only this user?

File: ncm-opennebula/src/main/perl/opennebula.pm [hunk 4]
```diff
@@ -305,23 +310,30 @@ sub enable_node
 # new pass
 sub change_oneadmin_passwd
```
Comment (stdweird) [inline ncm-opennebula/src/main/perl/opennebula.pm]: rename the method, has nothing to do with oneadmin anymore

File: ncm-opennebula/src/main/perl/opennebula.pm [hunk 5]
```diff
@@ -515,33 +530,62 @@ sub manage_users
...
-    %opts = $self->set_oned_file_opts();
+    my $oned_templ = $self->process_template($data, $service);
+    %opts = $self->set_file_opts();
     return if ! %opts;
-    my $fh = $oned_templ->filewriter($ONED_CONF_FILE, %opts);
+    my $fh = $oned_templ->filewriter($config_file, %opts);
 
     if (!defined($fh)) {
-        $self->error("Failed to render $ONED_CONF_FILE (".$oned_templ->{fail}."). Skipping");
+        $self->error("Failed to render $config_file (".$oned_templ->{fail}."). Skipping");
```
Comment (stdweird) [inline ncm-opennebula/src/main/perl/opennebula.pm]: mention the `$service` in the error message

File: ncm-opennebula/src/main/perl/opennebula.pm [hunk 6]
```diff
@@ -515,33 +530,62 @@ sub manage_users
...
+
+    my $passwd = {$user => $data};
+    my $trd = $self->process_template($passwd, "one_auth");
+    %opts = $self->set_file_opts();
+    return if ! %opts;
+
+    if ($user eq "oneadmin") {
+        $self->verbose("Writing $user auth file: $ONEADMIN_AUTH_FILE");
+        $fh = $trd->filewriter($ONEADMIN_AUTH_FILE, %opts);
+        die "Problem rendering $ONEADMIN_AUTH_FILE" if (!defined($fh));
```
Comment (stdweird) [inline ncm-opennebula/src/main/perl/opennebula.pm]: use same error logic as in `set_one_service_conf` (in particular no `die`)

File: ncm-opennebula/src/main/perl/opennebula.pm [hunk 7]
```diff
@@ -515,33 +530,62 @@ sub manage_users
...
+        $fh->close();
+    } elsif ($user eq "serveradmin") {
+        foreach my $service (@SERVERADMIN_AUTH_FILE) {
+            $auth_file = $SERVERADMIN_AUTH_DIR . $service;
+            $self->verbose("Writing $user auth file: $auth_file");
+            $fh = $trd->filewriter($auth_file, %opts);
+            die "Problem rendering $auth_file" if (!defined($fh));
+            $fh->close();
+        }
+    }
```
Comment (stdweird) [inline ncm-opennebula/src/main/perl/opennebula.pm]: log error "unsupported user" in the `else` block

File: ncm-opennebula/src/main/resources/one_auth.tt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+[% FOR pair IN one_auth.pairs -%]
+[%    pair.key %]:[% pair.value %]
+[% END -%]
```
Comment (stdweird) [inline ncm-opennebula/src/main/resources/one_auth.tt]: add newline

File: ncm-opennebula/src/main/resources/sunstone_level1.tt [hunk 9]
```diff
@@ -0,0 +1,5 @@
+[%-    FOR pair IN data.pairs -%]
+[%         FILTER indent -%]
+[%             data.pairs.first.key == pair.key ? "- " : "  " %]:[% pair.key %]: [% pair.value %]
+[%         END %]
+[%-    END %]
```
Comment (stdweird) [inline ncm-opennebula/src/main/resources/sunstone_level1.tt]: add newline

File: ncm-opennebula/src/main/resources/tests/regexps/datastore_shar/share_template [hunk 10]
```diff
@@ -0,0 +1,10 @@
+OpenNebula shared datastore full test
+---
+multiline
```
Comment (stdweird) [inline ncm-opennebula/src/main/resources/tests/regexps/datastore_shar/share_template]: you can remove `multiline` tag, it's the default

File: ncm-opennebula/src/main/resources/tests/regexps/datastore_shar/simple [hunk 11]
```diff
@@ -0,0 +1,10 @@
+OpenNebula share datastore test
+---
+multiline
```
Comment (stdweird) [inline ncm-opennebula/src/main/resources/tests/regexps/datastore_shar/simple]: same here

File: ncm-opennebula/src/main/resources/tests/regexps/sunstone/simple [hunk 12]
```diff
@@ -0,0 +1,34 @@
+OpenNebula sunstone_server test
+---
+multiline
```
Comment (stdweird) [inline ncm-opennebula/src/main/resources/tests/regexps/sunstone/simple]: and here

File: ncm-opennebula/src/main/resources/one_auth.tt (hunk 8 above)
Comment (stdweird) [inline ncm-opennebula/src/main/resources/one_auth.tt]: also add unittest

**Comment by stdweird** [comment]
@alvarosimon  looks good, some minor reamrks (try to factor out the filewriter error handling in private method?)

File: ncm-opennebula/src/main/pan/components/opennebula/schema.pan [hunk 13]
```diff
@@ -99,6 +99,15 @@ type opennebula_tm_mad_conf = {
 } = dict();
 
 @documentation{ 
+The following attributes define the default cost for Virtual Machines that don’t have a CPU or MEMORY cost.
```
Comment (stdweird) [inline ncm-opennebula/src/main/pan/components/opennebula/schema.pan]: the `'` in `don't` is not a `'` but some othe rfunky character (reported by @StephaneGerardVUB)

File: ncm-opennebula/src/main/perl/opennebula.pm (hunk 3 above)
Comment (alvarosimon) [inline ncm-opennebula/src/main/perl/opennebula.pm]: This is a special case, serveradmin and oneadmin are internal system users used by OpenNebula, so if the component detects this username in your tree it changes the pass by the new one. I will include this info into the documentation to be more clear.

**Comment by alvarosimon** [comment]
@stdweird  @StephaneGerardVUB thanks for your remarks, I will include these changes today

**Comment by alvarosimon** [comment]
Tested in altaria, the sunstone-server.conf is working fine

**Comment by StephaneGerardVUB** [comment]
I was able to create a shared datastore. It works, thanks.

**Comment by alvarosimon** [comment]
@StephaneGerardVUB great! :+1:

**Comment by jrha** [comment]
@alvarosimon is this ready? There are a couple of remarks from @stdweird that I am unsure whether you have dealt with.

**Comment by alvarosimon** [comment]
@jrha I think that the filewriter error refactoring and unit tests remarks are already addressed,  @stdweird are you happy with these changes?

**Comment by alvarosimon** [comment]
@jrha @stdweird ups I found an issue in this PR due to users passwd change, I will fix it, please do not merge this yet...

**Comment by stdweird** [comment]
retest this please

**Comment by alvarosimon** [comment]
@stdweird tested in altaria, now the passwords are not displayed into ncm logs even if *_auth files are changed

**Comment by stdweird** [comment]
lgtm

**Pull request merged**